File: examples/MatlabSparseMatrix/MatlabSparseMatrixExample.cxx
#include "rtkMatlabSparseMatrix.h"
#include "rtkThreeDCircularProjectionGeometry.h"
#include "rtkConstantImageSource.h"
#include "rtkJosephBackProjectionImageFilter.h"
#include "rtkStoreSparseMatrixSplatWeightMultiplication.h"
#include <vnl/vnl_sparse_matrix.h>
#include <fstream>
#include <itkImage.h>
#include <iostream>

int
main()
{
  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;

  constexpr unsigned int numberOfProjections = 10;
  constexpr unsigned int sid = 600;  // source to isocenter distance
  constexpr unsigned int sdd = 1200; // source to detector distance

  // Set up the geometry
  auto geometry = rtk::ThreeDCircularProjectionGeometry::New();
  for (unsigned int i = 0; i < numberOfProjections; ++i)
  {
    double angle = (360.0 * i) / numberOfProjections;
    geometry->AddProjectionInRadians(0, 0, angle * itk::Math::pi / 180.0, sid, sdd, 0, 0);
  }

  // Create projection images
  auto projectionsSource = rtk::ConstantImageSource<OutputImageType>::New();
  projectionsSource->SetSize(itk::MakeSize(512, 1, numberOfProjections));
  projectionsSource->SetSpacing(itk::MakeVector(1.0, 1.0, 1.0));
  projectionsSource->SetOrigin(itk::MakePoint(-256.0, 0.0, 0.0));
  projectionsSource->SetConstant(1.0);
  projectionsSource->Update();

  // Create volume
  auto volumeSource = rtk::ConstantImageSource<OutputImageType>::New();
  volumeSource->SetSize(itk::MakeSize(32, 32, 32));
  volumeSource->SetSpacing(itk::MakeVector(2.0, 2.0, 2.0));
  volumeSource->SetOrigin(itk::MakePoint(-32.0, -32.0, -32.0));
  volumeSource->SetConstant(0.0);
  volumeSource->Update();

  // Back-project with matrix capture
  using BackProjectionType = rtk::JosephBackProjectionImageFilter<
    OutputImageType,
    OutputImageType,
    rtk::Functor::InterpolationWeightMultiplicationBackProjection<OutputPixelType, OutputPixelType>,
    rtk::Functor::StoreSparseMatrixSplatWeightMultiplication<OutputPixelType, double, OutputPixelType>>;

  auto backProjection = BackProjectionType::New();
  backProjection->SetInput(volumeSource->GetOutput());
  backProjection->SetInput(1, projectionsSource->GetOutput());
  backProjection->SetGeometry(geometry);

  // Initialize and configure matrix capture
  backProjection->GetSplatWeightMultiplication().GetVnlSparseMatrix().resize(
    projectionsSource->GetOutput()->GetLargestPossibleRegion().GetNumberOfPixels(),
    volumeSource->GetOutput()->GetLargestPossibleRegion().GetNumberOfPixels());
  backProjection->GetSplatWeightMultiplication().SetProjectionsBuffer(
    projectionsSource->GetOutput()->GetBufferPointer());
  backProjection->GetSplatWeightMultiplication().SetVolumeBuffer(volumeSource->GetOutput()->GetBufferPointer());

  backProjection->Update();

  // Export to Matlab format
  vnl_sparse_matrix<double> & systemMatrix = backProjection->GetSplatWeightMultiplication().GetVnlSparseMatrix();
  auto matlabMatrix = rtk::MatlabSparseMatrix<OutputImageType>::New();
  matlabMatrix->SetMatrix(systemMatrix);
  matlabMatrix->SetOutput(volumeSource->GetOutput());

  std::ofstream outputFile("backprojection_matrix.mat", std::ios::binary);
  matlabMatrix->Save(outputFile);

  // Print matrix information
  matlabMatrix->Print();

  return 0;
}

File: examples/MatlabSparseMatrix/MatlabSparseMatrixExample.py
#!/usr/bin/env python

import itk
import numpy as np

# Define types
OutputPixelType = itk.F
Dimension = 3
OutputImageType = itk.Image[OutputPixelType, Dimension]

# Parameters
numberOfProjections = 10
sid = 600  # source to isocenter distance
sdd = 1200  # source to detector distance

# Set up the geometry
GeometryType = itk.ThreeDCircularProjectionGeometry
geometry = GeometryType.New()
for i in range(numberOfProjections):
    angle = (360.0 * i) / numberOfProjections
    geometry.AddProjectionInRadians(0, 0, angle * np.pi / 180.0, sid, sdd, 0, 0)

# Create projection images
projectionsSource = itk.ConstantImageSource[OutputImageType].New()
projectionsSource.SetSize([512, 1, numberOfProjections])
projectionsSource.SetSpacing([1.0, 1.0, 1.0])
projectionsSource.SetOrigin([-256.0, 0.0, 0.0])
projectionsSource.SetConstant(1.0)
projectionsSource.Update()

# Create volume
volumeSource = itk.ConstantImageSource[OutputImageType].New()
volumeSource.SetSize([32, 32, 32])
volumeSource.SetSpacing([2.0, 2.0, 2.0])
volumeSource.SetOrigin([-32.0, -32.0, -32.0])
volumeSource.SetConstant(0.0)
volumeSource.Update()

# Back-project with matrix capture
IWM = itk.InterpolationWeightMultiplicationBackProjection[
    OutputPixelType, OutputPixelType
]
SSM = itk.StoreSparseMatrixSplatWeightMultiplication[
    OutputPixelType, itk.D, OutputPixelType
]
backProjection = itk.JosephBackProjectionImageFilter[
    OutputImageType, OutputImageType, IWM, SSM
].New()
backProjection.SetInput(volumeSource.GetOutput())
backProjection.SetInput(1, projectionsSource.GetOutput())
backProjection.SetGeometry(geometry)

# Initialize and configure matrix capture
backProjection.GetSplatWeightMultiplication().Resize(
    projectionsSource.GetOutput().GetLargestPossibleRegion().GetNumberOfPixels(),
    volumeSource.GetOutput().GetLargestPossibleRegion().GetNumberOfPixels(),
)
backProjection.GetSplatWeightMultiplication().SetProjectionsBuffer(
    projectionsSource.GetOutput().GetBufferPointer()
)
backProjection.GetSplatWeightMultiplication().SetVolumeBuffer(
    volumeSource.GetOutput().GetBufferPointer()
)

backProjection.Update()

# Export to Matlab format
matlabMatrix = itk.MatlabSparseMatrix[OutputImageType].New()
matlabMatrix.SetMatrixFromFunctor(backProjection.GetSplatWeightMultiplication())
matlabMatrix.SetOutput(volumeSource.GetOutput())

matlabMatrix.Save("backprojection_matrix.mat")

# Print matrix information
matlabMatrix.Print()

print("Matrix successfully saved to backprojection_matrix.mat")
